core/src/lib.rs

```diff
@@ -129,45 +147,86 @@ impl Machine {
 
 	/// Loop stepping the machine, until it stops.
 	pub fn run(&mut self) -> Capture<ExitReason, Trap> {
+		let mut handler = SimpleInterpreterHandler::default();
 		loop {
-			match self.step() {
+			match self.step(&mut handler, &H160::default()) {
 				Ok(()) => (),
 				Err(res) => return res,
 			}
 		}
 	}
 
 	#[inline]
-	/// Step the machine, executing one opcode. It then returns.
-	pub fn step(&mut self) -> Result<(), Capture<ExitReason, Trap>> {
+	/// Step the machine, executing until exit or trap.
+	pub fn step<H: InterpreterHandler>(
+		&mut self,
+		handler: &mut H,
+		address: &H160,
+	) -> Result<(), Capture<ExitReason, Trap>> {
 		let position = *self
 			.position
 			.as_ref()
 			.map_err(|reason| Capture::Exit(reason.clone()))?;
-
-		match self.code.get(position).map(|v| Opcode(*v)) {
-			Some(opcode) => match eval(self, opcode, position) {
-				Control::Continue(p) => {
-					self.position = Ok(position + p);
-					Ok(())
-				}
-				Control::Exit(e) => {
-					self.position = Err(e.clone());
-					Err(Capture::Exit(e))
-				}
-				Control::Jump(p) => {
-					self.position = Ok(p);
-					Ok(())
-				}
-				Control::Trap(opcode) => {
-					self.position = Ok(position + 1);
-					Err(Capture::Trap(opcode))
-				}
-			},
-			None => {
-				self.position = Err(ExitSucceed::Stopped.into());
-				Err(Capture::Exit(ExitSucceed::Stopped.into()))
+		match eval(self, position, handler, address) {
+			Control::Continue(_) | Control::Jump(_) => {
+				unreachable!("must not be here, eval computes branches");
```

Comment:
Might be worth changing the return type of `eval` to only include `Trap` and `Exit` to avoid needing this `unreachable`. At the very least, putting this at the bottom of the block might be ever so slightly better performance-wise (I think matches are usually simply checked top-down).

Reply:
Not sure how to express subset of enum in return type. Regarding performance - note that here we return rather rarely, as most of the program will be executed inside the loop. Anyway, reordered branches.